packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2462,6 +2463,16 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     }
   }
 
+  void _updateCaretRectIfNeeded() {
+    if (_hasInputConnection) {
+      final TextPosition currentTextPosition = TextPosition(offset: renderEditable.selection!.baseOffset);
```

Comment:
Addressed [here](https://github.com/flutter/flutter/pull/77608/commits/a652c0324adefdae2d03fb88dd2d87e0065f3a2c).

Reply:
Looks like we are skipping reporting the caret rect to the engine when there's no caret. Does the engine need to know that the caret is gone (instead of using the out-of-date caret rect that was previously sent to it)?

Reply:
The purpose of caret rect is to position the IME candidate popup. The popup is displayed after long pressing a key (or possibly while typing with some languages). So if there is any selection it will be replaced by a character before the popup is displayed. I'm not aware of a way to get the IME popup visible without caret present. And should that ever happen, last caret position seems to me like a good enough fallback.

Reply:
Wouldn't the accent popup show up at the wrong location if you:
1. change the selection to somewhere in the document that's far away from the current caret location (the framework won't send anything in this case because it's a non-collapsed selection).
2. long press a key to trigger the popup

Reply:
Long pressing key first clears the selection replacing it with pressed character, moving caret to new position. I don't think this would cause any issues. The way this works is that the popup shows the moment character would normally start repeating (after a long key press).

Reply:
The communication between the text input plugin and the framework is asynchronous. When the software keyboard clears the selection and changes the caret's position, the text input plugin needs to query the new caret coordinates from the framework. There seems to be a chance that the framework can't respond in time, resulting in the outdated caret coordinates being reported in the `firstRect` method?

Reply:
I'm not entirely sure what do you mean by software keyboard? if you mean mean the selection popup, that is shown as response to previous keypress (that would clear the selection). The time between the initial key press and showing popup (when long pressing the key) is about a second, plenty of time for framework to notify plugin of cursor location.

Reply:
ah sorry forgot this is macOS & hardware keyboard, I meant the key event. Sounds good.